[5.0] docs(treelist, gantt): Add Column Menu documentation (#1711)
telerik/blazor-docs: components/treelist/columns/menu.md

---
title: Column Menu
page_title: TreeList - Column Menu

> description: Use the Column Menu for the TreeList to show a menu that allows you to perform high-level column customization 

Reviewer: ```suggestion
description: Use the Column Menu for the TreeList to show a menu that allows you to perform high-level column customization.
```

> description: Use the Column Menu for the TreeList to show a menu that allows you to perform high-level column customization.

Reviewer 2: What are "high-level customizations" (which ones are "low-level"?) and why do we need to label them if we list them afterwards? And if we don't list them, like here, maybe we should.

In other words, don't introduce unnecessary terminology.

slug: treelist-column-menu
tags: telerik,blazor,treelist,column,columns,menu
published: True
position: 20
---

# Column Menu

The TreeList enables you to show a menu with quick actions for its columns. The Column Menu enables you to perform high-level customization like [sorting]({%slug treelist-sorting%}), [filtering]({%slug treelist-filtering%}), [showing or hiding]({%slug treelist-columns-visible%}) columns and [freezing or unfreezing]({%slug treelist-columns-frozen%}) them.

>caption In this article:
* [Basics](#basics)
* [Features](#features)
    * [Column Chooser](#column-chooser)
    * [Filtering](#filtering)

>     * [Groupable](#groupable)

Reviewer 2: This doesn't link anywhere

    * [Frozen Columns](#frozen-columns)
    * [Sections](#sections)
    * [Sorting](#sorting)
    * [Reorderable](#reorderable)
* [Examples](#examples)
* [Notes](#notes)

## Basics

To enable the Column Menu, set the `ShowColumnMenu` parameter of the `<TelerikTreeList>` tag to `true`. This will enable the menu for each column of the TreeList.

> To disable the Column Menu for a specific column in the TreeList, set the `ShowColumnMenu` parameter of the column to `false`.

Reviewer 2: You can show this in the example too.

                        Name = $"first level child {j} of {i}",
                        HireDate = DateTime.Now.AddDays(-currId),
                        Salary = 2000 + i * 200
                    });

                for (int k = 1; k < 3; k++)
                {
                    int nestedId = currId * 1000 + k;
                    data.Add(new Employee
                        {
                            Id = nestedId,
                            ParentId = currId,
                            Name = $"second level child {k} of {i} and {currId}",
                            HireDate = DateTime.Now.AddMinutes(-nestedId),
                            Salary = 1000 + i * 200
                        }); ;
                }
            }
        }

        return await Task.FromResult(data);
    }

    protected override async Task OnInitializedAsync()
    {
        TreeListData = await GetTreeListData();
    }

    public class Employee
    {
        public int Id { get; set; }
        public int? ParentId { get; set; }
        public string Name { get; set; }
        public DateTime HireDate { get; set; }
        public int Salary { get; set; }
    }
}
````

## Features

> To control the common features of the Column Menu, use the `<TreeListColumnMenuSettings>` tag, nested inside the `<TreeListSettings>` tag. By default, all Column Menu features are enabled.

Reviewer: ```suggestion
To control the features of the Column Menu, use the `<TreeListColumnMenuSettings>` tag, nested inside the `<TreeListSettings>` tag. By default, all Column Menu features are enabled.
```

### Column Chooser

> The Column Chooser in the Column Menu allows you to toggle the visibility of TreeLis columns. By default, all columns are visible under the **Columns** section of the Column Menu. To expand the menu, click the **Columns** item.

Reviewer 2: ```suggestion
The Column Chooser in the Column Menu allows you to toggle the visibility of TreeList columns. By default, all columns are visible under the **Columns** section of the Column Menu. To expand the menu, click the **Columns** item.
```

* To disable the column chooser, set the `ShowColumnChooser` parameter of the `<TreeLisColumnMenuSettings>` to `false`.
* To hide a column from the Column Chooser, set the `VisibleInColumnChooser` property of the column to `false`.

### Filtering

To control whether filtering is possible from the Column Menu, set the `FilterMode` parameter of the `TreeListColumnMenuSettings` tag to a member of the `ColumnMenuFilterMode` enum:

* `None`—disables the filtering from the Column Menu. This is the recommended option if you use the [`FilterRow` mode]({%slug treelist-filter-row%}).
* `FilterMenu`—enables filtering from a filter menu.

### Frozen Columns

To disable the locking and unlocking of a column from the Column Menu, set the `Lockable` parameter of the column to `false`.

### Sorting

To remove the sorting option from the Column Menu, set the `Sortable` parameter of the `TreeListColumnMenuSettings` tag to `false`.

### Reorderable

To allow column reordering from the Column Menu, set the `Reorderable` parameter of the `TreeListColumnMenuSettings` tag to `true`.

> ### Sections

Reviewer 2: Column Sections may be better. Just "Sections" is too general.

The TreeList Column Menu lets you group the columns in the [Column Chooser](#column-chooser) into different sections:

1. Use the `TreeListColumnMenuChooser` tag (child to the `TreeListColumnMenuSettings`).

1. Add the [Template]({%slug treelist-templates-column-chooser%}) tag.

1. Provide a `TreeListColumnMenuChooserGroup` which is a collection of the columns that will be in the section. To render a title for the section, use the `Title` parameter.

1. Use the `TreeListColumnMenuChooserItem` to denote the columns that belong to the group.

    * You must set the `ColumnId` parameter of the `TreeListColumnMenuChooserItem` to the value of the `Id` parameter of the corresponding TreeList Column.
    
    * If you set the `Title` parameter of the `TreeListColumnMenuChooserItem`, it will override the value of the `Title` parameter of the corresponding TreeList Column. 

> ## Examples 

Reviewer 2: The two child headings doesn't make it clear what is the difference between the two examples. What is the difference between "features" and "configuration"? Better use a single example with everything, and there will be no need to split the features to "basic" and "advanced".

(I see that it's like that in the Grid, which doesn't guarantee it's the best option.)

        <TreeListColumn Field="@nameof(Employee.Position)" VisibleInColumnChooser="false"/>
        <TreeListColumn Field="@nameof(Employee.HireDate)" />
    </TreeListColumns>
</TelerikTreeList>

@code {

    List<Employee> TreeListData { get; set; }

    protected override void OnInitialized()
    {
        TreeListData = new List<Employee>();

        for (int i = 1; i <= 9; i++)
        {
            TreeListData.Add(new Employee()
                {
                    Id = i,
                    FirstName = "First " + i,
                    LastName = "Last " + i,
                    Position = i <= 3 ? "Team Lead" : "Software Engineer",
                    HireDate = DateTime.Now.AddDays(-i).Date
                });
        }

        base.OnInitialized();
    }

    public class Employee
    {
        public int Id { get; set; }
        public int? ParentId { get; set; }
        public string FirstName { get; set; }
        public string LastName { get; set; }
        public string Position { get; set; }
        public DateTime HireDate { get; set; }
    }
}
````

## Notes

> * Applying settings like `Filterable="false"`, `Sortable="false"`, `Lockable="false"` to a TreeList column takes precedence over the common settings applied in the `<TreeListColumnMenuSettings>` and disables the above-mentioned functionalities for the corresponding column.

Reviewer: ```suggestion
* Settings applied to a TreeList column take precedence over the settings applied to the Column Menu through the `<TreeListColumnMenuSettings>` tag. For example, if you set `Filterable="false"` to a TreeList column and `Filterable="true"` to a TreeList Column Menu, the filtering functionality will be disabled.
```

---
title: Column Menu
page_title: TreeList - Column Menu

> description: Use the Column Menu for the TreeList

Reviewer: The length of the description must be around 150 symbols.

slug: treelist-column-menu
tags: telerik,blazor,treelist,column,columns,menu
published: True
position: 20
---

# Column Menu

> 

Reviewer: The comments from the `gantt/gantt-tree/menu.md` file apply here as well.
